capytaine/bem/engines.py

```diff
@@ -315,3 +327,144 @@ def get_col_func(j):
+        """
+        this_block = matrix
+        for index in path:
+            this_block = this_block.all_blocks[index, index]
+        return this_block
+
+    def build_matrices(self,
+                       mesh1, mesh2, free_surface, water_depth, wavenumber, green_function,
+                       _rec_depth=1):
+        """Recursively builds a hierarchical matrix between mesh1 and mesh2,
+        and precomputes some of the quantities needed for the preconditioner.
+
+        Same arguments as :func:`BasicMatrixEngine.build_matrices`, plus
+        the properties of the hierarchical tree: leaf_order and path_to_leaf
+
+        :code:`_rec_depth` keeps track of the recursion depth only for pretty log printing.
+        """
+        # Build the matrices using the method of the parent class
+        S, K = super().build_matrices(mesh1, mesh2, free_surface, water_depth,
+                                      wavenumber, green_function,
+                                      _rec_depth=_rec_depth)
+
+        leaf_order = mesh1.leaf_order
+        path_to_leaf = mesh1.path_to_leaf
+
+        n = len(leaf_order)
+        N = K.shape[0]
+
+        # Navigate to the diagonal blocks and compute their LU decompositions
+        DLU = []
+        diag_shapes = []
+        for leaf in leaf_order:
+            # Navigate to the block containing the one we need
+            # (one layer above in the dendrogram)
+            upper_block = self.access_block_by_path(K, path_to_leaf[leaf][:-1])
+            # find the local index in the full path
+            ind = path_to_leaf[leaf][-1]
+            # compute the LU decomposition and add to the list
+            DLU.append(lu_factor(upper_block.all_blocks[ind, ind]))
+            diag_shapes.append(upper_block.all_blocks[ind, ind].shape[0])
+
+        # Build the restriction and precompute its multiplication by K
+        R = np.zeros((n, N), dtype=complex)
+        RA = np.zeros((n, N), dtype=complex)
+        for ii in range(n):
+            row_slice = slice(sum(diag_shapes[:ii]), sum(diag_shapes[:ii+1]))
+            R[ii, row_slice] = 1
+            # Compute the multiplication using only the relevant slices of K
+            # The slices are found by navigating the tree
+            #RA[ii, :] = self.slice_rmatvec(R[ii, :], ii)
+            Aloc = K
+            v = R[ii, :]
+            va = np.zeros(N, dtype=complex)
+            free = [0, N]
+
+            for lvl, jj in enumerate(path_to_leaf[leaf_order[ii]]):
+
+                Nrows = Aloc.all_blocks[jj, jj].shape[0]
+
+                if jj==0:
+                    v = v[:Nrows]
+                    #try-except block needed because of different function
+                    #definitions for rmatmul in np.ndarrays, blockmatrix and lowrank
+                    try:
+                        w = Aloc.all_blocks[0,1].__rmatmul__(v)
+                    except:
+                        w = Aloc.all_blocks[0,1].rmatvec(v)
```

Comment:
If `__rmatmul__` is missing somewhere (`BlockMatrix` in `block.py`?), maybe we should just implement it there to avoid the complexity here.